File: src/Systems/Pascal/chain_of_responsibility.pas
program ChainOfResponsibilityExample;

{$mode objfpc}{$H+}

uses
  SysUtils;

type
  THandler = class
  private
    FName: string;
    FLimit: Integer;
    FNext: THandler;
  public
    constructor Create(const AName: string; ALimit: Integer; ANext: THandler);
    destructor Destroy; override;
    function Handle(Amount: Integer; var Visited: string): string;
  end;

constructor THandler.Create(const AName: string; ALimit: Integer; ANext: THandler);
begin
  inherited Create;
  FName := AName;
  FLimit := ALimit;
  FNext := ANext;
end;

destructor THandler.Destroy;
begin
  FNext.Free;
  inherited Destroy;
end;

function THandler.Handle(Amount: Integer; var Visited: string): string;
begin
  if Visited <> '' then
    Visited := Visited + '>';
  Visited := Visited + FName;

  if Amount <= FLimit then
    Exit(FName);

  if FNext = nil then
    raise Exception.Create('No handler accepted the request');

  Result := FNext.Handle(Amount, Visited);
end;

var
  Root: THandler;
  Visited, Handled: string;
  Amount: Integer;
begin
  Amount := 250;
  Root := THandler.Create('faq', 50,
    THandler.Create('billing', 500,
      THandler.Create('escalation', MaxInt, nil)));
  try
    Visited := '';
    Handled := Root.Handle(Amount, Visited);
    WriteLn(Format('visited=%s;handled=%s;result=refund(%d)',
      [Visited, Handled, Amount]));
  finally
    Root.Free;
  end;
end.

File: src/Systems/Fortran/chain_of_responsibility.f90
program chain_of_responsibility
  implicit none
  type :: handler
     character(len=10) :: name
     integer :: limit
  end type handler

  type(handler), dimension(3) :: chain
  character(len=32) :: visited
  character(len=10) :: handled
  integer :: amount, i

  chain(1) = handler('faq', 50)
  chain(2) = handler('billing', 500)
  chain(3) = handler('escalation', huge(1))
  amount = 250
  visited = ''
  handled = ''

  do i = 1, size(chain)
     if (len_trim(visited) == 0) then
        visited = trim(chain(i)%name)
     else
        visited = trim(visited) // '>' // trim(chain(i)%name)
     end if

     if (amount <= chain(i)%limit) then
        handled = trim(chain(i)%name)
        exit
     end if
  end do

  if (trim(handled) /= 'billing') error stop 'unexpected handler'
  write(*,'(A,I0,A)') 'visited=' // trim(visited) // ';handled=' // trim(handled) // ';result=refund(', amount, ')'
end program chain_of_responsibility
